io/src/lib.rs

```diff
@@ -0,0 +1,345 @@
+    /// block side length in the canvas. The block side length can be
+    /// obtained by [`NFTPixelboardState::BlockSideLength`].
+    /// * NFT `rectangle` mustn't collide with already minted one.
+    /// * `painting` length must equal a pixel count in an NFT
+    /// (which can be calculated by multiplying a [width](`Rectangle::width`)
+    /// & [height](`Rectangle::height`) from `rectangle`).
+    /// * [`msg::source()`] must have enough tokens to buy all free pixels that
+    /// `rectangle` will occupy. An enough number of tokens can be calculated by
+    /// multiplying `rectangle` area and a price of a free pixel. The area can
+    /// be calculated by multiplying a [width](`Rectangle::width`) &
+    /// [height](`Rectangle::height`) from `rectangle`. The price of a free
+    /// pixel can be obtained by [`NFTPixelboardState::PixelPrice`].
+    ///
+    /// On success, returns [`NFTPixelboardEvent::Minted`].
+    ///
+    /// [`msg::source()`]: gstd::msg::source
+    Mint {
+        rectangle: Rectangle,
+        token_metadata: TokenMetadata,
+        /// A painting that'll be displayed in a place of an NFT on a pixelboard
+        /// after a successful minting.
+        painting: Vec<Color>,
+    },
+
+    /// Buys an NFT minted on a pixelboard.
+    ///
+    /// Transfers a purchased NFT from a pixelboard to [`msg::source()`].
+    ///
+    /// **Note:** If [`msg::source()`] has enough tokens to pay a resale
+    /// commission but not the entire token, then the commission will still be
+    /// withdrawn from its account.
+    ///
+    /// # Requirements
+    /// * An NFT must be minted on a pixelboard.
+    /// * An NFT must be for sale. This can be found out by
+    /// [`NFTPixelboardState::TokenInfo`]. See also the documentation of
+    /// [`TokenInfo#structfield.pixel_price`].
+    /// * [`msg::source()`] must have enough tokens to buy all pixels that a
+    /// token occupies. This can be found out by
+    /// [`NFTPixelboardState::TokenInfo`]. See also the documentation of
+    /// [`TokenInfo#structfield.pixel_price`].
+    ///
+    /// On success, returns [`NFTPixelboardEvent::Bought`].
+    ///
+    /// [`msg::source()`]: gstd::msg::source
+    Buy(TokenId),
+
+    /// Puts up for sale an NFT minted on a pixelboard.
+    ///
+    /// Transfers a minted NFT to a pixelboard program.
+    ///
+    /// **Note:** A commission is included in each token resale, so a seller
+    /// will receive not all tokens but tokens with a commission deduction. A
+    /// commission percentage can be obtained by
+    /// [`NFTPixelboardState::CommissionPercentage`].
+    ///
+    /// # Requirements
+    /// * An NFT must be minted on a pixelboard.
+    /// * [`msg::source()`](gstd::msg::source) must be the owner of an NFT.
+    /// * An NFT mustn't be for sale.
+    /// * `pixel_price` mustn't be more than [`MAX_PIXEL_PRICE`].
+    ///
+    /// On success, returns [`NFTPixelboardEvent::ForSale`].
+    PutUpForSale {
```

Comment:
May be it is worth adding the ability to remove from the sale and return NFT to the owner or update the price?